### PR TITLE
Address updater feedback

### PR DESCRIPTION
**Pull request by quanganhdo**
Task/Issue URL: https://app.asana.com/0/1201048563534612/1209026627874986/f
Tech Design URL:
CC:

**Description**:

Addresses some updater feedback:

- "Install Now" menu item does nothing
- Incorrect Release Notes screen state

Also adds more logging to help track down the edge cases (non-producible on my machine, but some internal users kept running into them) 

**Optional E2E tests**:
- [ ] Run PIR E2E tests
	Check this to run the Personal Information Removal end to end tests. If updating CCF, or any PIR related code, tick this.

**Steps to test this PR**:
1. Download this test build: https://staticcdn.duckduckgo.com/macos-desktop-browser/a81f9679-e41d-404d-81a6-fd4a461d82f7/testbuilds/5ab32a1/duckduckgo-1.120.0.333.dmg
2. Try different scenarios; use this as reference: https://app.asana.com/0/1199230911884351/1208882660080993/f
3. Run this Terminal command: `log show --info --predicate 'subsystem == "Updates"' --style compact --last 24h > ~/Downloads/updater.log`
4. Give the log file a quick skim; all major update steps should be there.

**Definition of Done**:

* [ ] Does this PR satisfy our [Definition of Done](https://app.asana.com/0/1202500774821704/1207634633537039/f)?

---
###### Internal references:
[Pull Request Review Checklist](https://app.asana.com/0/1202500774821704/1203764234894239/f)
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
[Pull Request Documentation](https://app.asana.com/0/1202500774821704/1204012835277482/f)

### CODE REVIEW

File: DuckDuckGo/Updates/ReleaseNotesTabExtension.swift [hunk 1]
```diff
@@ -145,14 +145,15 @@ extension ReleaseNotesValues {
...
+        let lastUpdate = UInt((updateController.lastUpdateCheckDate ?? Date()).timeIntervalSince1970)
 
-        guard let updateController, let latestUpdate = updateController.latestUpdate else {
-            self.init(status: updateController?.updateProgress.toStatus ?? .loaded,
+        guard let latestUpdate = updateController.latestUpdate else {
+            self.init(status: updateController.updateProgress.toStatus,
                       currentVersion: currentVersion,
-                      lastUpdate: lastUpdate)
+                      lastUpdate: lastUpdate,
+                      automaticUpdate: updateController.areAutomaticUpdatesEnabled)
```
Comment (quanganhdo) [inline DuckDuckGo/Updates/ReleaseNotesTabExtension.swift:156]: This happens when Release Notes screen is displayed but the appcast hasn't been loaded yet. We get both incorrect status (updateReady), wrong button text (Restart to Update / Update DuckDuckGo), and the button doesn't even work.

File: DuckDuckGo/Updates/ReleaseNotesTabExtension.swift [hunk 2]
```diff
@@ -198,7 +199,7 @@ private extension UpdateCycleProgress {
         case .downloadDidStart, .downloading: return .updateDownloading
         case .extractionDidStart, .extracting, .readyToInstallAndRelaunch, .installationDidStart, .installing: return .updatePreparing
         case .updaterError: return .updateError
-        case .updateCycleNotStarted, .updateCycleDone: return .updateReady
+        case .updateCycleNotStarted, .updateCycleDone: return .loaded
```
Comment (quanganhdo) [inline DuckDuckGo/Updates/ReleaseNotesTabExtension.swift:202]: These should've been defaulted to `loaded` as we only want `updateReady` if we truly have a pending update.

File: DuckDuckGo/Updates/ReleaseNotesUserScript.swift [hunk 3]
```diff
@@ -77,7 +77,10 @@ final class ReleaseNotesUserScript: NSObject, Subfeature {
             return
         }
 
-        let updateController = Application.appDelegate.updateController
+        guard let updateController = Application.appDelegate.updateController else {
+            return
+        }
```
Comment (quanganhdo) [inline DuckDuckGo/Updates/ReleaseNotesUserScript.swift:82]: Just to satisfy the compiler as AppDelegate.updateController is defined as `UpdateController!`

File: DuckDuckGo/Updates/UpdateController.swift [hunk 4]
```diff
@@ -274,30 +274,50 @@ extension UpdateController: SPUUpdaterDelegate {
...
 
     func updater(_ updater: SPUUpdater, willInstallUpdate item: SUAppcastItem) {
-        Logger.updates.log("Updater will install update: \(item.displayVersionString)(\(item.versionString))")
+        Logger.updates.log("Updater will install update: \(item.displayVersionString, privacy: .public)(\(item.versionString, privacy: .public))")
+    }
+
+    func updater(_ updater: SPUUpdater, willInstallUpdateOnQuit item: SUAppcastItem, immediateInstallationBlock immediateInstallHandler: @escaping () -> Void) -> Bool {
+        Logger.updates.log("Updater will install update on quit: \(item.displayVersionString, privacy: .public)(\(item.versionString, privacy: .public))")
+        userDriver?.configureResumeBlock(immediateInstallHandler)
+        return true
```
Comment (quanganhdo) [inline DuckDuckGo/Updates/UpdateController.swift:292]: This delegate method isn't called in my testing scenarios, but when it does in prod, let's make sure we use the `immediateInstallHandler` to set up our resume block for Install Now.

File: DuckDuckGo/Updates/UpdateUserDriver.swift [hunk 5]
```diff
@@ -122,21 +133,27 @@ final class UpdateUserDriver: NSObject, SPUUserDriver {
...
+        Logger.updates.log("Updater showed update found: (userInitiated:  \(state.userInitiated, privacy: .public), stage: \(state.stage.rawValue, privacy: .public))")
+        sparkleUpdateState = state
+
         if appcastItem.isInformationOnlyUpdate {
+            Logger.updates.log("Updater dismissed due to information only update")
             reply(.dismiss)
         }
 
-        onSkipping = { reply(.skip) }
+        onDismiss = { reply(.dismiss) }
```
Comment (quanganhdo) [inline DuckDuckGo/Updates/UpdateUserDriver.swift:149]: Skip can ignore the update, which we don't want.

File: DuckDuckGo/Updates/UpdateUserDriver.swift [hunk 6]
```diff
@@ -184,19 +201,26 @@ final class UpdateUserDriver: NSObject, SPUUserDriver {
...
         updateProgress = .updateCycleDone
     }
 
     func showInstallingUpdate(withApplicationTerminated applicationTerminated: Bool, retryTerminatingApplication: @escaping () -> Void) {
         updateProgress = .installationDidStart
+        
+        if !applicationTerminated {
+            Logger.updates.log("Updater re-sent a quit event")
+            retryTerminatingApplication()
+        }
```
Comment (quanganhdo) [inline DuckDuckGo/Updates/UpdateUserDriver.swift:223]: Not strictly necessary, but there seems to be no harm in doing this.

File: DuckDuckGo/Updates/ReleaseNotesTabExtension.swift [hunk 7]
```diff
@@ -194,11 +195,11 @@ private extension Update {
 private extension UpdateCycleProgress {
     var toStatus: ReleaseNotesValues.Status {
         switch self {
-        case .updateCycleDidStart: return .loading
+        case .updateCycleNotStarted, .updateCycleDidStart: return .loading
         case .downloadDidStart, .downloading: return .updateDownloading
         case .extractionDidStart, .extracting, .readyToInstallAndRelaunch, .installationDidStart, .installing: return .updatePreparing
         case .updaterError: return .updateError
-        case .updateCycleNotStarted, .updateCycleDone: return .updateReady
+        case .updateCycleDone: return .loaded
```
Comment (quanganhdo) [inline DuckDuckGo/Updates/ReleaseNotesTabExtension.swift:202]: `updateReady` should never be returned here unless we're truly ready to proceed with an update.

**Review by jotaemepereira — APPROVED** [review]
@quanganhdo I went through the different scenarios on: https://app.asana.com/0/1199230911884351/1208882660080993/f, and everything work as expected. Let’s merge this into the release branch 👍🏼
